Source/JavaScriptCore/runtime/JSInternalPromiseConstructor.cpp

```diff
@@ -64,4 +75,65 @@ JSInternalPromiseConstructor::JSInternalPromiseConstructor(VM& vm, FunctionExecu
 {
 }
 
+// InternalPromise.internalAll(array)
+// This function is intended to be used in the JSC internals.
+// The implementation should take care not to perform the user observable / trappable operations.
+// 1. Don't use for-of and iterables. This function only accepts the dense array of the promises.
+// 2. Don't look up this.constructor / @@species. Always construct the plain InternalPromise object.
+JSC_DEFINE_HOST_FUNCTION(internalPromiseConstructorFuncInternalAll, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    auto* promise = JSInternalPromise::create(vm, globalObject->internalPromiseStructure());
+
+    auto callReject = [&]() -> void {
+        Exception* exception = scope.exception();
+        ASSERT(exception);
+        scope.clearException();
+        scope.release();
+        promise->reject(vm, globalObject, exception);
+    };
+
+    JSValue arrayValue = callFrame->argument(0);
+    JSArray* array = jsDynamicCast<JSArray*>(arrayValue);
+    ASSERT(array);
+
+    uint64_t length = array->length();
+    if (!length) {
+        JSArray* values = JSArray::tryCreate(vm, globalObject->arrayStructureForIndexingTypeDuringAllocation(ArrayWithContiguous), 0);
+        if (!values) [[unlikely]] {
+            throwOutOfMemoryError(globalObject, scope);
+            callReject();
+            return JSValue::encode(promise);
+        }
+        scope.release();
+        // Use fulfill instead of resolve to avoid looking up the then property.
+        promise->fulfill(vm, globalObject, values);
+        return JSValue::encode(promise);
+    }
+
+    JSArray* values = JSArray::tryCreate(vm, globalObject->arrayStructureForIndexingTypeDuringAllocation(ArrayWithContiguous), length);
```

Comment:
Since `performPromiseThenWithInternalMicrotask` could take a bit, I wonder if it should use `MarkedArgumentBuffer` instead and then only create the array at the very end once all operations complete?